### ISSUE TITLE
Missing required POST parameters

### ISSUE THREAD

**Issue by Ateean** [plotly/arduino-api#52]
Hi folks,

I'm new to plot.ly and ran in a couple issue from the beginning. I'm using Arduino IDE 1.0.5 (I was using 1.6 before the issues but reverted to the old rev to make sure the IDE was not causing it.)

I am using the simple_ethernet example with my credentials (user name, API key and tokens, MAC, IP). I get the following messages on the serial window:
-----------------------------------------------
... Initializing ethernet
... Done initializing ethernet
... Attempting to connect to plotly's REST servers
... Connected to plotly's REST servers
... Sending HTTP Post to plotly
POST /clientresp HTTP/1.1
Host: plot.ly:80
User-Agent: Arduino/0.6.0
Content-Length: 337

version=2.3&origin=plot&platform=arduino&un=XXXXXX&key=e1edy6jq48&args=[{"y": [], "x": [], "type": "scatter", "stream": {"token": "n99wrrs60x", "maxpoints": 100}}, {"y": [], "x": [], "type": "scatter", "stream": {"token": "klu8ps0id2", "maxpoints": 100}}]&kwargs={"fileopt": "extend", "filename": "Internet_test", "world_readable": true}
... Sent message, waiting for plotly's response...
HTTP/1.1 200 OK
Content-Type: application/json
Date: Sat, 02 Jan 2016 17:37:51 GMT
Server: nginx/1.4.6 (Ubuntu)
Set-Cookie: csrftoken=pS28DJHm6X2dt912F26m3csKwy2ScM4x; expires=Mon, 02-Jan-2017 17:37:51 GMT; Max-Age=31622400; Path=/; secure
Set-Cookie: anoncsrf=8r2crCQENsmRPpAQo43e4pNgeKrxMIBV; expires=Sun, 03-Jan-2016 17:37:51 GMT; httponly; Max-Age=86400; Path=/
Vary: Accept-Encoding
Vary: Cookie
X-Frame-Options: SAMEORIGIN
Content-Length: 138
Connection: keep-alive

{"url": "", "message": "", "warning": "", "filename": "", "error": "Missing required POST parameters: platform un key origin args kwargs"}HTTP/1.1 400 BAD_REQUEST
Content-Length: 0
Connection: Close

... Error initializing stream, aborting. Try again or get in touch with Chris at chris@plot.ly
----------------------------------
I don't see what are the missing POST parameters. They all seem to be there. When I compare the string sent to plot.ly to the other ones found in the "issues" page, they look pretty similar.

Here's the code I'm using:
----------------------------------
#include <Ethernet.h>
#include <SPI.h>
#include "plotly_streaming_ethernet.h"

// Sign up to plotly here: https://plot.ly
// View your API key and streamtokens here: https://plot.ly/settings
#define nTraces 2
// View your tokens here: https://plot.ly/settings
// Supply as many tokens as data traces
// e.g. if you want to ploty A0 and A1 vs time, supply two tokens
char *tokens[nTraces] = {"n99wrrs60x", "klu8ps0id2"};
// arguments: username, api key, streaming token, filename
plotly graph = plotly("XXXXXX", "e1edy6jq48", tokens, "Internet_test", nTraces);

byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
byte my_ip[] = { XX, X, X, XX }; // google will tell you: "public ip address"

void startEthernet(){
    Serial.println("... Initializing ethernet");
    if(Ethernet.begin(mac) == 0){
        Serial.println("... Failed to configure Ethernet using DHCP");
        // no point in carrying on, so do nothing forevermore:
        // try to congifure using IP address instead of DHCP:
        Ethernet.begin(mac, my_ip);
    }
    Serial.println("... Done initializing ethernet");
    delay(1000);
}

void setup() {

  // Open serial communications and wait for port to open:
  Serial.begin(9600);
  while (!Serial) {
    ; // wait for serial port to connect. Needed for Leonardo only
  }

  startEthernet();

  graph.fileopt="extend"; // See the "Usage" section in https://github.com/plotly/arduino-api for details
  bool success;
  success = graph.init();
  if(!success){while(true){}}
  graph.openStream();
}

unsigned long x;
int y;

void loop() {
  graph.plot(millis(), analogRead(A0), tokens[0]);
  graph.plot(millis(), analogRead(A1), tokens[1]);

  delay(50);

}
-------------------------------
Any help would be greatly appreciated.

Thanks!
Etienne

**Comment by DedeHai** [comment]
Same problem here.

**Comment by Ateean** [comment]
Hi DedeHai

Have you found a fix for this issue?

**Comment by DedeHai** [comment]
nope. they seem to have changed something in the API but there is no documentation about it. I tried to modify the HTTP header, no luck. plot.ly just refuses the initialization. I even wrote an email to plot.ly, no response.

**Comment by suprnrdy** [comment]
Take a look at issue #41, it seemed to solve this problem too.

**Comment by ergose** [comment]
I'm getting similar. I tried the fixes mentioned on #41 to no avail. I've omitted the cookies and styles below to keep clean. Here's what I'm getting back...
```
HTTP/1.1 200 OK
Server: nginx
Date: Sat, 27 May 2017 19:33:42 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Authorization, Cookie, Accept-Language
X-Frame-Options: SAMEORIGIN
x-xss-protection: 1; mode=block
Content-Language: en
Set-Cookie: plotly_sess_pr= [--- omitted ---] ; Domain=.plot.ly; expires=Sat, 10-Jun-2017 19:33:42 GMT; httponly; Max-Age=1209600; Path=/; secure
Set-Cookie: plotly_csrf_pr= [--- omitted ---]; Domain=.plot.ly; expires=Mon, 28-May-2018 19:33:42 GMT; Max-Age=31622400; Path=/; secure
Set-Cookie: plotly_anoncsrf_pr= [--- omitted ---]; Domain=.plot.ly; expires=Sun, 28-May-2017 19:33:42 GMT; httponly; Max-Age=86400; Path=/
Via: 1.1 google
Set-Cookie: [--- omitted ---] ; path=/; HttpOnly; expires=Sun, 28-May-2017 19:33:42 GMT

8a
{"url": "", "message": "", "warning": "", "filename": "", "error": "Missing required POST parameters: platform un key origin args kwargs"}
0

HTTP/1.0 400 Bad Request
Content-Type: text/html; charset=UTF-8
Referrer-Policy: no-referrer
Content-Length: 1555
Date: Sat, 27 May 2017 19:33:42 GMT

<!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 400 (Bad Request)!!1</title>
  <style>
   [--- omitted ---]
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>400.</b> <ins>That's an error.</ins>
  <p>Your client has issued a malformed or illegal request.  <ins>That's all we know.</ins>
```

I checked the REST API docs here: https://plot.ly/rest/

I also checked plotly_streaming_cc3000.cpp and it looks like the required params are being set. I copy-pasted fresh the API Key, two tokens from the settings page, and set the filename to a short one-word name, so below should be getting the correct info. Ideas? Is a codefix needed here? (I'm dusty on my C.)
```
   // Start printing querystring body
    print_(F("version=2.3&origin=plot&platform=arduino&un="));
    print_(username_);
    print_(F("&key="));
    print_(api_key_);
    print_(F("&args=["));
    // print a trace for each token supplied
    for(int i=0; i<nTraces_; i++){
        print_(F("{\"y\": [], \"x\": [], \"type\": \"scatter\", \"stream\": {\"token\": \""));
        print_(stream_tokens_[i]);
        print_(F("\", \"maxpoints\": "));
        print_(maxpoints);
        print_(F("}}"));
        if(nTraces_ > 1 && i != nTraces_-1){
            print_(F(", "));
        }
    }
    print_(F("]&kwargs={\"fileopt\": \""));
    print_(fileopt);
    print_(F("\", \"filename\": \""));
    print_(filename_);
    print_(F("\", \"world_readable\": "));
    if(world_readable){
        print_("true");
    } else{
        print_("false");
    }
    print_(F("}"));
```
